consensus/event_loop.go

```diff
@@ -0,0 +1,854 @@
+
+	return fc.currentView, nil
+}
+
+// ProcessTipBlock must only be called when the server has accepted a new block, connected it
+// to the tip of the blockchain, and determined that the block is safe to vote on. Given such a
+// block, this function resets internal state and schedules the next crank timer and timeout
+// timer.
+//
+// Expected params:
+//   - tip: the current uncommitted tip of the blockchain, with the validator set at that block height
+//   - safeBlocks: an unordered slice of blocks including the committed tip, the uncommitted tip,
+//     all ancestors of the uncommitted tip that are safe to extend from, and all blocks from forks
+//     that are safe to extend from. This function does not validate the collection of blocks. It
+//     expects the server to know and decide what blocks are safe to extend from.
+func (fc *FastHotStuffEventLoop) ProcessTipBlock(tip BlockWithValidators, safeBlocks []BlockWithValidators) error {
+	// Grab the event loop's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the event loop is running
+	if fc.status != eventLoopStatusRunning {
+		return errors.New("FastHotStuffEventLoop.ProcessTipBlock: Event loop is not running")
+	}
+
+	// Validate the safe blocks and validator sets, and store them
+	if err := fc.storeBlocks(tip, safeBlocks); err != nil {
+		return errors.Wrap(err, "FastHotStuffEventLoop.ProcessTipBlock: ")
+	}
+
+	// We track the current view here so we know which view to time out on later on.
+	fc.currentView = fc.tip.block.GetView() + 1
+
+	// Reset QC construction status for the current view
+	fc.hasCrankTimerRunForCurrentView = false
+	fc.hasConstructedQCInCurrentView = false
+
+	// Evict all stale votes and timeouts
+	fc.evictStaleVotesAndTimeouts()
+
+	// Signal the server that we can vote for the block. The server will decide whether to construct and
+	// broadcast the vote.
+	fc.Events <- &FastHotStuffEvent{
+		EventType:      FastHotStuffEventTypeVote,
+		TipBlockHash:   fc.tip.block.GetBlockHash(),
+		TipBlockHeight: fc.tip.block.GetHeight(),
+		View:           fc.tip.block.GetView(),
+	}
+
+	// Schedule the next crank timer and timeout scheduled tasks
+	fc.resetScheduledTasks()
+
+	return nil
+}
+
+// setSafeBlocks is a helper function that validates the provided blocks, validator sets, and stores them.
+// It must be called while holding the event loop's lock.
+func (fc *FastHotStuffEventLoop) storeBlocks(tip BlockWithValidators, safeBlocks []BlockWithValidators) error {
+	// Do a basic integrity check on the tip block and validator set
+	if !isProperlyFormedBlock(tip.Block) || !isProperlyFormedValidatorSet(tip.Validators) {
+		return errors.New("Invalid tip block or validator set")
+	}
+
+	// Do a basic integrity check on the blocks and validator sets
+	hasMalformedInput := collections.Any(safeBlocks, func(block BlockWithValidators) bool {
+		return !isProperlyFormedBlock(block.Block) || !isProperlyFormedValidatorSet(block.Validators)
+	})
```

Comment:
Nit: Prefer pulling embedded functions out of function calls like this as follows:

```
	checkMalformedInput := func(block BlockWithValidators) bool {
		return !isProperlyFormedBlock(block.Block) || !isProperlyFormedValidatorSet(block.Validators)
	}

	// Do a basic integrity check on the blocks and validator sets
	hasMalformedInput := collections.Any(safeBlocks, checkMalformedInput)
```